### PR TITLE
DDF-3441 Adds support for custom and predefined themes in Intrigue

### PR DESCRIPTION
**Pull request by andrewkfiedler**
#### What does this PR do?
     - Refactors less to better support a wide variety of possible themes.
     - Adds support for predefined themes as well as custom themes.
     - Adds color picker input for use in theming

#### Who is reviewing it? 
@andrewkfiedler
@bdeining
@rzwiefel 

@djblue 
@adimka 

#### How should this be tested? (List steps with links to updated documentation)
As an Admin, go to the UI config and select a theme for the UI.  Verify that a user gets that theme upon first visiting the UI.  Select custom for the theme and verify that the custom theme is used upon first visiting the UI.

As a User, verify that you can override your initial theme once within the UI by going to Theme under Settings.  

Verify that themes persist.

#### What are the relevant tickets?
https://codice.atlassian.net/browse/DDF-3441

#### Screenshots (if appropriate)
screen shot 2017-11-15 at 2 00 12 am
screen shot 2017-11-15 at 1 59 41 am
screen shot 2017-11-15 at 1 59 29 am
screen shot 2017-11-15 at 1 59 12 am

### CODE REVIEW

File: catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/metatype/org.codice.ddf.catalog.ui.config.xml [hunk 1]
```diff
@@ -302,6 +302,94 @@
...
+        <AD id="customWarningColor"
+            name="Warning Color"
+            description="Warning Color"
+            type="String"
+            default="#c89600"
+            required="true"/>
+
+        <AD id="customFavoriteColor"
+            name="Favorite Color"
+            description="Favor Color"
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/metatype/org.codice.ddf.catalog.ui.config.xml]: nit : `Favorite` color?

File: catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/metatype/org.codice.ddf.catalog.ui.config.xml [hunk 2]
```diff
@@ -302,6 +302,94 @@
...
+            type="String"
+            default="#d1d179"
+            required="true"/>
+
+        <AD id="customBackgroundNavigation"
+            name="Background Navigation"
+            description="Background Navigation"
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/metatype/org.codice.ddf.catalog.ui.config.xml]: nit : Might be worthwhile to add `Color` to these descriptions for clarity

File: catalog/ui/catalog-ui-search/src/main/webapp/component/map-info/map-info.less [hunk 3]
```diff
@@ -1,16 +1,15 @@
...
-    color: @text-color;
     font-size: @minimumFontSize;
     position: absolute;
     left: 0px;
     bottom: 0px;
     text-align: left;
     padding: @minimumSpacing;
-    background: rgba(0, 0, 0, .7);
-    max-width: 50%; //text-shadow: 0px 0px 5px @dropdown-background-color, 0px 0px 5px @dropdown-background-color, 0px 0px 5px @dropdown-background-color, 0px 0px 5px @dropdown-background-color,  0px 0px 5px @dropdown-background-color,  0px 0px 5px @dropdown-background-color;
+    max-width: 50%; //text-shadow: 0px 0px 5px @backgroundDropdown, 0px 0px 5px @backgroundDropdown, 0px 0px 5px @backgroundDropdown, 0px 0px 5px @backgroundDropdown,  0px 0px 5px @backgroundDropdown,  0px 0px 5px @backgroundDropdown;
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/map-info/map-info.less]: nit : do we need the` text-shadow:` comment?

File: catalog/ui/catalog-ui-search/src/main/webapp/js/model/Theme.js [hunk 4]
```diff
@@ -60,30 +72,39 @@ var colorModes = {
...
     initialize: function() {
     },
+    getCustomColorNames: function() {
+        return colorVariables;
+    },
     getTheme: function(){
         var theme = this.toJSON();
-        return _.defaults(theme, spacingModes[theme.spacingMode], colorModes[theme.colorMode]);
+
+        return _.defaults(theme, spacingModes[theme.spacingMode]/* , colorModes[theme.colorMode] */);
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/webapp/js/model/Theme.js]: Do we need the comment here?

**Comment by glenhein** [comment]
If a user selects a custom theme (under the Theme menu), and only fills in one color, then the screen becomes unreadable. There is no way to return to a sane color scheme. Can we have the custom colors default the current colors?

**Comment by glenhein** [comment]
Can we have the custom colors in the admin config default to the basic ddf colors? It looks like the metatype has the defaults, but it looks like org.codice.ddf.catalog.ui.config.config doesn't set the correct variable names for the colors.

File: catalog/ui/search-ui-app/src/main/resources/etc/org.codice.ddf.catalog.ui.config.config [hunk 5]
```diff
@@ -61,3 +61,14 @@ uiName="Intrigue"
 showRelevanceScores=B"false"
 relevancePrecision=I"5"
 showLogo=B"false"
+
+backgroundColor="#424249"
```
Comment (glenhein) [inline catalog/ui/search-ui-app/src/main/resources/etc/org.codice.ddf.catalog.ui.config.config]: These names don't match the names in the metatype xml.

**Comment by glenhein** [comment]
If I manually fill-in the custom colors in the admin config, then some of the user-specific custom color pickers are pre-populated, but not all of them.
screen shot 2017-11-15 at 8 57 52 am

**Comment by glenhein** [comment]
Hero looks good. The problems I experienced earlier have been resolved. I did not run itests because of the time required. Please watch for CI to run the itests.

**Comment by bdeining** [comment]
build now

File: catalog/ui/catalog-ui-search/src/main/webapp/component/input/textarea/input-textarea.view.js [hunk 6]
```diff
@@ -16,31 +16,60 @@
...
+    },
+    onAttach: function() {
+        InputView.prototype.onAttach.call(this);
+        this.hasOverflowed();
     },
     getCurrentValue: function () {
         return this.$el.find('textarea').val();
     },
-    isOverflowing: function () {
+    hasOverflowed: function() {
```
Comment (millerw8) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/input/textarea/input-textarea.view.js]: Is this related?  Or just another clean-up that is sneaking in?

Comment (andrewkfiedler) [reply catalog/ui/catalog-ui-search/src/main/webapp/component/input/textarea/input-textarea.view.js]: I had to go touch up the styling around notes and couldn't help myself.  I noticed we always show the more button even when there isn't anything more to show.  So I went and fixed it.  

There are more changes we should make to notes in the future to make it more consistent with the rest of the UI, but we can save those for a future PR.

**Comment by andrewkfiedler** [comment]
build now

**Comment by bdeining** [comment]
build now

File: catalog/ui/catalog-ui-search/src/main/webapp/js/workers/worker.less.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+                return [];
+            } else if (tagName === 'link') {
+                return [];
+            }
+        }
+    };
+})();
+
+const Less = require('less');
+const variableRegex = '/@(.*:[^;]*)/g';
```
Comment (rzwiefel) [inline catalog/ui/catalog-ui-search/src/main/webapp/js/workers/worker.less.ts]: Eventually it might be nice for these to be refined since they seem a little broad as is.

**Comment by bdeining** [comment]
Looks good.  As a nit, the theme is not selected by default, which seems strange.
screen shot 2017-11-28 at 12 46 29 pm

**Comment by bdeining** [comment]
@andrewkfiedler if i set my preference to `Sea` and reload the UI, it will load as `Dark` and then switch over to the `Sea` theme.

**Comment by andrewkfiedler** [comment]
@bdeining that's expected behavior.  It starts with the base theme and then calculates whatever the user has set and applies it.

Strange that the theme didn't appear selected by default for you though.  I kind of want to look into that.

**Comment by bdeining** [comment]
This may be expected behavior, but to watch the theme change during the loading process seems unintuitive and might be perceived as a bug by users.

File: catalog/ui/catalog-ui-search/src/main/webapp/js/model/Theme.js [hunk 8]
```diff
@@ -32,6 +47,25 @@ function trimVariables(variables){
...
+        return false;
+    }
+    
+    var image = document.createElement("img");
+    image.style.color = "rgb(0, 0, 0)";
+    image.style.color = stringToTest;
+    if (image.style.color !== "rgb(0, 0, 0)") { return true; }
+    image.style.color = "rgb(255, 255, 255)";
+    image.style.color = stringToTest;
+    return image.style.color !== "rgb(255, 255, 255)";
```
Comment (Variadicism) [inline catalog/ui/catalog-ui-search/src/main/webapp/js/model/Theme.js]: Again, I'm curious about this pattern.

File: catalog/ui/catalog-ui-search/src/main/webapp/component/input/color/input-color.view.js [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+            return false;
+        }
+        
+        var image = document.createElement("img");
+        image.style.color = "rgb(0, 0, 0)";
+        image.style.color = stringToTest;
+        if (image.style.color !== "rgb(0, 0, 0)") { return true; }
+        image.style.color = "rgb(255, 255, 255)";
+        image.style.color = stringToTest;
+        return image.style.color !== "rgb(255, 255, 255)";
```
[the remark this answers is not in the document]
Comment (andrewkfiedler) [reply catalog/ui/catalog-ui-search/src/main/webapp/component/input/color/input-color.view.js]: It verifies that the color is valid.  If the `image.style.color` doesn't update, that means it isn't valid.

**Comment by andrewkfiedler** [comment]
build now

**Comment by andrewkfiedler** [comment]
build now

**Comment by andrewkfiedler** [comment]
build now
